src/Ocean/HydrostaticBoussinesq/HydrostaticBoussinesqModel.jl

```diff
@@ -633,9 +635,9 @@ function update_auxiliary_state_gradient!(
     reverse_indefinite_stack_integral!(dg, m, Q, A, t, elems) # top -> bottom
 
     # project w(z=0) down the stack
-    # Need to be consistent with vars_state_auxiliary
-    # A[1] = w, A[3] = wz0
-    copy_stack_field_down!(dg, m, A, 1, 3, elems)
+    index_w = varsindex(vars_state_auxiliary(m, FT), :w)[1]
+    index_wz0 = varsindex(vars_state_auxiliary(m, FT), :wz0)[1]
+    copy_stack_field_down!(dg, m, A, index_w, index_wz0, elems)
```

Comment:
Another option is to just broadcast and avoid the `copy_stack_field_down!`
```suggestion
    wz0 = reshape(A.wz0, Nq^2, Nq, 1, nhorz_elem, nvert_elem)
    w = reshape(A.w, Nq^2, Nq, 1, nhorz_elem, nvert_elem)
    wz0 = @view w[:, end:end, :, :, end:end]
```

Reply:
Actually doesn't work. results in the error
```
ERROR: LoadError: ArgumentError: an array of type `CuArrays.CuArray` shares memory with another argument and must
make a preventative copy of itself in order to maintain consistent semantics,
but `copy(A)` returns a new array of type `CuArrays.CuArray{Float32,3,Nothing}`. To fix, implement:
    `Base.unaliascopy(A::CuArrays.CuArray)::typeof(A)`
Stacktrace:
 [1] _unaliascopy(::CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}}, ::CuArrays.CuArray{Float32,3,Nothing}) at ./abstractarray.jl:1174
 [2] unaliascopy(::CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}}) at ./abstractarray.jl:1172
 [3] unaliascopy(::SubArray{Float32,3,CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}}},false}) at ./subarray.jl:100
 [4] unaliascopy(::Base.ReshapedArray{Float32,5,SubArray{Float32,3,CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}}},false},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}}) at ./reshapedarray.jl:214
 [5] unaliascopy(::SubArray{Float32,5,Base.ReshapedArray{Float32,5,SubArray{Float32,3,CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}}},false},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}},Base.Slice{Base.OneTo{Int64}},UnitRange{Int64}},false}) at ./subarray.jl:100
 [6] unalias at ./abstractarray.jl:1155 [inlined]
 [7] broadcast_unalias at ./broadcast.jl:884 [inlined]
 [8] preprocess at ./broadcast.jl:891 [inlined]
 [9] preprocess_args at ./broadcast.jl:894 [inlined]
 [10] preprocess at ./broadcast.jl:890 [inlined]
 [11] copyto! at ./broadcast.jl:907 [inlined]
 [12] copyto! at ./broadcast.jl:863 [inlined]
 [13] materialize!(::Base.ReshapedArray{Float32,5,SubArray{Float32,3,CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}}},false},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{5},Nothing,typeof(identity),Tuple{SubArray{Float32,5,Base.ReshapedArray{Float32,5,SubArray{Float32,3,CuArrays.CuArray{Float32,3,CuArrays.CuArray{Float32,3,Nothing}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}}},false},Tuple{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}}},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64},Base.Slice{Base.OneTo{Int64}},Base.Slice{Base.OneTo{Int64}},UnitRange{Int64}},false}}}) at ./broadcast.jl:822
 [14] top-level scope at /home/jekozdon/codes/CLIMA/try.jl:36
 [15] include at ./boot.jl:328 [inlined]
 [16] include_relative(::Module, ::String) at ./loading.jl:1105
 [17] include(::Module, ::String) at ./Base.jl:31
 [18] include(::String) at ./client.jl:424
 [19] top-level scope at REPL[1]:1
in expression starting at /home/jekozdon/codes/CLIMA/try.jl:13
```

when tested with this code

```julia
using MPI

using ClimateMachine
using ClimateMachine.MPIStateArrays
using ClimateMachine.VariableTemplates: @vars, varsindex, varsize
using StaticArrays

ClimateMachine.init()
const ArrayType = ClimateMachine.array_type()
const mpicomm = MPI.COMM_WORLD

let                                                                                                                                                                                                                                                                                                                                                                             V = @vars begin
        w::Float32                                                                                                                                                                                                                                                                                                                                                                  junk_1::SVector{3, Float32}
        wz0::Float32                                                                                                                                                                                                                                                                                                                                                            end
    nvert_elem = 10                                                                                                                                                                                                                                                                                                                                                             nhorz_elem = 100
    nelem = nvert_elem * (nhorz_elem+10) # +10 for some ghost elems
    realelems = 1:(nvert_elem * nhorz_elem)                                                                                                                                                                                                                                                                                                                                     nstate = varsize(V)
    Nq = 5
    Np = Nq^3
    A = MPIStateArray{Float32, V}(
        mpicomm,
        ArrayType,
        Np,
        nstate,
        nelem;
        realelems = realelems,
    )
    wz0 = reshape(A.wz0, Nq^2, Nq, 1, nhorz_elem, nvert_elem)
    w = reshape(A.w, Nq^2, Nq, 1, nhorz_elem, nvert_elem)
    wz0 .= @view w[:, end:end, :, :, end:end]
end
nothing
```

Reply:
yep, this what I think we should do long term. but @leios has been trying to get broadcasts of reshaped MPIstatearrays to work. Not sure if the @view in L640 would circumvent his issues.

Reply:
I have a PR up to do this (#1071). It needs a lot of polish and testing, though. Just ran out of time to finish it.

Reply:
I don't think that #1071 should be needed here, since that `A.w` and `A.wz0` pulls out a view into the `CuArray` data. (There is no more `MPIStateArray` at that point.)